src/triage/component/catwalk/predictors.py

```diff
@@ -151,75 +147,100 @@ def _write_to_db(
-                        prediction.as_of_date,
-                        prediction.score,
-                        prediction.label_value,
-                        prediction.rank_abs,
-                        prediction.rank_pct,
-                        prediction.matrix_uuid,
-                        prediction.test_label_timespan
-                    ])
-                f.seek(0)
-                postgres_copy.copy_from(f, Prediction, self.db_engine, format='csv')
+                )
+            )
         else:
             logging.info('as_of_date not found as part of matrix index, using '
                          'matrix metadata end_time as as_of_date')
-            temp_df = pandas.DataFrame({'score': predictions})
-            rankings_abs = temp_df['score'].rank(method='dense', ascending=False)
-            rankings_pct = temp_df['score'].rank(method='dense', ascending=False, pct=True)
-            for entity_id, score, label, rank_abs, rank_pct in zip(
-                matrix_store.matrix.index,
-                predictions,
-                labels,
-                rankings_abs,
-                rankings_pct
-            ):
-                db_objects.append(Prediction(
+            db_objects_generator = (
+                Prediction(
                     model_id=int(model_id),
                     entity_id=int(entity_id),
                     as_of_date=matrix_store.metadata['end_time'],
                     score=round(float(score), 10),
                     label_value=int(label) if not math.isnan(label) else None,
-                    rank_abs=int(rank_abs),
-                    rank_pct=round(float(rank_pct), 10),
                     matrix_uuid=matrix_store.uuid,
                     test_label_timespan=test_label_timespan,
                     **misc_db_parameters
-                ))
-
-            session.bulk_save_objects(db_objects)
-            session.commit()
-            session.close()
+                )
+                for entity_id, score, label in zip(
+                    matrix_store.matrix.index,
+                    predictions,
+                    labels,
+                )
+            )
+        logging.info(
+            'Beginning COPY of Prediction database objects for model %s, matrix %s',
+            model_id,
+            matrix_store.uuid
+        )
+        save_db_objects(self.db_engine, db_objects_generator)
+        logging.info(
+            'Completed COPY of Prediction database objects for model %s, matrix %s',
+            model_id,
+            matrix_store.uuid
+        )
+        logging.info(
+            'Beginning ranking of new Predictions for model %s, matrix %s',
+            model_id,
+            matrix_store.uuid
+        )
+        with self.db_engine.begin() as conn:
+            conn.execute('''
+                with ranks as (
+                    select
+                        entity_id,
+                        as_of_date,
+                        dense_rank() over (
```

Comment:
I think we want row_number() here instead of dense_rank() to make sure we don't have ties in ranks.